src/debugger/src/cli_debugger.cpp

```diff
@@ -0,0 +1,154 @@
+		 */
+		str = "./tdb --" + str;
+		cmd_attributes_gen(str);
+	} while (!run_command());
+}
+
+bool CliDebugger::run_command() {
+	auto parsed_args = command_parser.parse(argc, argv);
+	if (parsed_args["run"].as<bool>()) {
+		debugger_core->run();
+		return true;
+	}
+	if (parsed_args["step"].as<bool>()) {
+		debugger_core->step();
+		return true;
+	}
+	if (parsed_args["help"].as<bool>()) {
+		cout << command_parser.help() << "\n";
+		return true;
+	}
+	if (parsed_args["peek"].as<uint32_t>() > 0) {
+		auto code_map = debugger_core->peek(parsed_args["peek"].as<uint32_t>());
+		for (auto i : code_map) {
+			cout << hex << i.first << ": " << i.second << "\n";
+		}
+		return true;
+	}
+	if (parsed_args["bp-set"].as<string>() != "NULL") {
+		auto code = debugger_core->set_breakpoint(
+		    string_to_hex(parsed_args["bp-set"].as<string>()));
+		if (!code)
+			cout << "bp already set!"
+			     << "\n";
+		return code;
+	}
+	if (parsed_args["bp-remove"].as<string>() != "NULL") {
+		auto code = debugger_core->remove_breakpoint(
+		    string_to_hex(parsed_args["bp-remove"].as<string>()));
+		if (!code)
+			cout << "bp not present!"
+			     << "\n";
+		return code;
+	}
+	if (parsed_args["bp_ticks-set"].as<ClockCycles>() > 0) {
+		auto code = debugger_core->set_tick_breakpoint(
+		    parsed_args["bp_ticks-set"].as<ClockCycles>());
+		if (!code)
+			cout << "bp already set!"
+			     << "\n";
+		return code;
+	}
+	if (parsed_args["bp_ticks-remove"].as<ClockCycles>() > 0) {
+		auto code = debugger_core->remove_tick_breakpoint(
+		    parsed_args["bp_ticks-remove"].as<ClockCycles>());
+		if (!code)
+			cout << "bp not present!"
+			     << "\n";
+		return code;
+	}
+	if (parsed_args["bp_cycles-set"].as<ClockCycles>() > 0) {
+		auto code = debugger_core->set_cycle_breakpoint(
+		    parsed_args["bp_ticks-set"].as<ClockCycles>());
+		if (!code)
+			cout << "bp already set!"
+			     << "\n";
+		return code;
+	}
+	if (parsed_args["bp_cycles-remove"].as<ClockCycles>() > 0) {
+		auto code = debugger_core->remove_cycle_breakpoint(
+		    parsed_args["bp_ticks-remove"].as<ClockCycles>());
+		if (!code)
+			cout << "bp not present!"
+			     << "\n";
+		return code;
+	}
+	if (parsed_args["bp-view"].as<bool>()) {
+		auto instr_bp = debugger_core->get_breakpoints();
+		cout << "The instructions breakpoints are: ";
+		for (auto i : instr_bp) {
+			cout << hex << i << " ";
+		}
+		cout << "\n";
+	}
+	if (parsed_args["bp_ticks-view"].as<bool>()) {
+		auto tick_bp = debugger_core->get_tick_breakpoints();
+		cout << "The tick breakpoints are: ";
+		for (auto i : tick_bp) {
+			cout << i << " ";
+		}
+		cout << "\n";
+	}
+	if (parsed_args["bp_cycles-view"].as<bool>()) {
+		auto cycles_bp = debugger_core->get_cycle_breakpoints();
+		cout << "The cycle breakpoints are: ";
+		for (auto i : cycles_bp) {
+			cout << i << " ";
+		}
+		cout << "\n";
+	}
+
+	return true;
+}
+
+void CliDebugger::cmd_attributes_gen(std::string command) {
```

Comment:
This can return a tuple of `(int argc, char **argv)`

With C++17 structured bindings, you can unpack them and declare on the caller side in a single line too.

```cpp
auto [argc, argv] = cmd_attributes_gen(command);
```
Handy!

Also, perhaps like `generate_argv_from_string` is a better name?

Reply:
`cmd_string_to_argv` would as well, your preference.